### PR TITLE
[6391]Caching with multithreading

### PR DESCRIPTION
**Pull request by KarolJagodzinski**
**Ticket**: [BAU-6391](https://bugzilla.stxnext.pl/show_bug.cgi?id=6391)

Added decorator for caching data from functions
Added tests

### CODE REVIEW

**Comment by KarolJagodzinski** [comment]
By the way. Changed seats in views for my convenience

File: src/presence_analyzer/utils.py [hunk 1]
```diff
@@ -19,6 +21,35 @@
 log = logging.getLogger(__name__)  # pylint: disable=invalid-name
 
 
+CACHE = {}
+
+
+def memoize(duration):
+    """
+    Decorator to caching data. Set duration in minutes.
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Should be in seconds as the ticket says.

File: src/presence_analyzer/tests.py [hunk 2]
```diff
@@ -234,6 +234,20 @@ def test_get_data_xml(self):
...
         self.assertEqual(data[176]['name'], 'Adrian K.')
 
+    def test_memoize(self):
+        """
+        Test memorize data and updating if expired.
+        """
+        utils.CACHE = {}
+        utils.get_data()
+        utils.get_data_xml()
+        self.assertTrue('get_data' in utils.CACHE.keys())
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: Use `assertIn` - https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertIn

Also no need to call `.keys()`

File: src/presence_analyzer/tests.py [hunk 3]
```diff
@@ -234,6 +234,20 @@ def test_get_data_xml(self):
...
+        self.assertTrue('value' and 'time' in utils.CACHE['get_data'].keys())
+        utils.CACHE['get_data']['time'] = 0
+        utils.get_data()
+        self.assertFalse(utils.CACHE['get_data']['time'] == 0)
+        self.assertTrue('get_data' and 'get_data_xml' in utils.CACHE.keys())
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: This doesn't work as I think you want it to. This is what you did with added parenthesis: 
`('get_data') and ('get_data_xml' in utils.CACHE)`. 
The first will be always `True` even it doesn't exist in the dictionary.

File: src/presence_analyzer/tests.py [hunk 4]
```diff
@@ -234,6 +234,20 @@ def test_get_data_xml(self):
         self.assertTrue(data[26]['avatar'].startswith('https://host:443/'))
         self.assertEqual(data[176]['name'], 'Adrian K.')
 
+    def test_memoize(self):
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: Missing a case showing that the data that is being returned is, indeed, from the cache.

File: src/presence_analyzer/utils.py [hunk 5]
```diff
@@ -19,6 +21,35 @@
...
+            key = function.__name__
+            now = time()
+
+            with lock:
+                if key in CACHE and CACHE[key]['time'] > now:
+                    return CACHE[key]['value']
+
+                result = function(*args, **kwargs)
+                CACHE[key] = {
+                    'time': now + duration*60,
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Missing spaces around the operator.

File: src/presence_analyzer/tests.py [hunk 6]
```diff
@@ -234,6 +237,25 @@ def test_get_data_xml(self):
...
+        """
+        Test memorize data and updating if expired.
+        """
+        utils.get_data()
+        self.assertIn('get_data', utils.CACHE)
+        self.assertIn('value', utils.CACHE['get_data'])
+        self.assertIn('time', utils.CACHE['get_data'])
+        utils.CACHE['get_data']['time'] = 0
+        utils.get_data()
+        self.assertFalse(utils.CACHE['get_data']['time'] == 0)
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: Use `assertNotEqual`

File: src/presence_analyzer/utils.py [hunk 7]
```diff
@@ -35,6 +66,7 @@ def inner(*args, **kwargs):
     return inner
 
 
+@memoize(5)
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: `600`

File: src/presence_analyzer/utils.py [hunk 8]
```diff
@@ -74,6 +106,7 @@ def get_data():
     return data
 
 
+@memoize(5)
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Ditto

**Pull request merged**